ufl/algorithms/remove_component_tensors.py

```diff
@@ -0,0 +1,117 @@
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+
+from ufl.classes import ComponentTensor, Form, Index, MultiIndex, Zero
+from ufl.corealg.map_dag import map_expr_dag
+from ufl.corealg.multifunction import MultiFunction, memoized_handler
+
+
+class IndexReplacer(MultiFunction):
+    """Replace Indices."""
+
+    def __init__(self, fimap: dict):
+        """Initialise.
+
+        Args:
+           fimap: map for index replacements.
+
+        """
+        MultiFunction.__init__(self)
+        self.fimap = fimap
+        self._object_cache = {}
+
+    expr = MultiFunction.reuse_if_untouched
+
+    @memoized_handler
+    def zero(self, o):
+        """Handle Zero."""
+        free_indices = []
+        index_dimensions = []
+        for i, d in zip(o.ufl_free_indices, o.ufl_index_dimensions):
+            k = Index(i)
+            j = self.fimap.get(k, k)
+            if isinstance(j, Index):
+                free_indices.append(j.count())
+                index_dimensions.append(d)
+        return Zero(
+            shape=o.ufl_shape,
+            free_indices=tuple(free_indices),
+            index_dimensions=tuple(index_dimensions),
+        )
+
+    @memoized_handler
+    def multi_index(self, o):
+        """Handle MultiIndex."""
+        return MultiIndex(tuple(self.fimap.get(i, i) for i in o.indices()))
+
+
+class IndexRemover(MultiFunction):
+    """Remove Indexed."""
+
+    def __init__(self):
+        """Initialise."""
+        MultiFunction.__init__(self)
+        self._object_cache = {}
+
+    expr = MultiFunction.reuse_if_untouched
+
+    @memoized_handler
+    def _unary_operator(self, o):
+        """Simplify UnaryOperator(Zero)."""
+        (operand,) = o.ufl_operands
+        f = map_expr_dag(self, operand)
+        if isinstance(f, Zero):
+            return Zero(
+                shape=o.ufl_shape,
+                free_indices=o.ufl_free_indices,
+                index_dimensions=o.ufl_index_dimensions,
+            )
+        if f is operand:
+            # Reuse if untouched
+            return o
+        return o._ufl_expr_reconstruct_(f)
+
+    @memoized_handler
+    def indexed(self, o):
+        """Simplify Indexed."""
+        o1, i1 = o.ufl_operands
+        if isinstance(o1, ComponentTensor):
+            # Simplify Indexed ComponentTensor
+            o2, i2 = o1.ufl_operands
+            # Replace inner indices first
+            v = map_expr_dag(self, o2)
+            # Replace outer indices
+            assert len(i2) == len(i1)
+            fimap = dict(zip(i2, i1))
+            rule = IndexReplacer(fimap)
+            return map_expr_dag(rule, v)
+
+        expr = map_expr_dag(self, o1)
+        if expr is o1:
+            # Reuse if untouched
+            return o
+        return o._ufl_expr_reconstruct_(expr, i1)
+
+    reference_grad = _unary_operator
+    reference_value = _unary_operator
+
+
+def remove_component_tensors(o):
+    """Remove component tensors."""
```

Comment:
Could we use `map_integrand_dags` here? @wence- pointed out the existence of this function to me last fall, and it cover way more logic than just form and Expr.